### PR TITLE
Rework how existing LXD PIDs are validated/discovered

### PR DESCRIPTION
**Pull request by simondeziel**
Change how LXD's `daemon.start` detects existing LXD processes left behind.

1. Optimize how `pgrep` is used to avoid needing to invoke it twice, once for regular daemons and another for debug ones.
2. Tweak `pgrep` invocation to only report PIDs that have an EUID of `0`. This ensures we don't accidentally get nested LXD PIDs because those should be in a different userns (for `unpriv`)
3. Once we have a candidate PID, make sure it's from a snap (`SNAP_NAME=lxd` line in `/proc/$pid/environ`)
4. Now that we have a probable LXD PID, make sure it is really a leftover process and not a nested LXD from a `privileged` container.

This last step involves writing short random data through the PID's mount namespace (`/proc/${pid}/root/var/snap/lxd/common/lxd/.validate`) and comparing it with our own view of the mount namespace by accessing the same file but directly (`/var/snap/lxd/common/lxd/.validate`).

### CODE REVIEW

File: snapcraft/commands/daemon.start [hunk 1]
```diff
@@ -620,7 +620,8 @@ fi
 # LXD
 ## Check for existing LXDs
 for pid in $(pgrep -f "lxd --logfile" ; pgrep -f "lxd.debug --logfile"); do
-    grep -qF "SNAP_NAME=lxd" "/proc/${pid}/environ" || continue
+    # Cheap confirmation that we likely have a LXD PID
+    grep -qxFm1 "SNAP_NAME=lxd" "/proc/${pid}/environ" || continue
```
Comment (tomponline) [inline snapcraft/commands/daemon.start]: Are there long form arguments we can use here? Either that or an extension of the commit message/comment to explain how this is an improvement/cheaper?

Comment (simondeziel) [reply snapcraft/commands/daemon.start]: While they all have long form versions, I think switching them all would make it harder to understand to some as most of those options I've only ever seen used in their short form.

I'm assuming the `-m1` bit is the one needing clarification here which I'm happy to expand as indeed it's not commonly used and the long form is quite clear (`--max-count=1`):
>        -m NUM, --max-count=NUM
>              Stop  reading  a  file  after  NUM matching lines.

Comment (tomponline) [reply snapcraft/commands/daemon.start]: actually it was most of them :)

Comment (tomponline) [reply snapcraft/commands/daemon.start]: but the use of --max-count makes sense to me now

File: snapcraft/commands/daemon.start [hunk 2]
```diff
@@ -619,7 +619,7 @@ fi
 
 # LXD
 ## Check for existing LXDs
-for pid in $(pgrep -f "lxd(\.debug)? --logfile"); do
+for pid in $(pgrep --euid 0 --full "lxd(\.debug)? --logfile"); do
```
Comment (tomponline) [inline snapcraft/commands/daemon.start:622]: in the example in your commit, why is server_pid 1293, but this command doesn't match that process?

Comment (tomponline) [reply snapcraft/commands/daemon.start:622]: Also why not --uid?

File: snapcraft/commands/daemon.start (hunk 1 above)
Comment (simondeziel) [reply snapcraft/commands/daemon.start]: I've now expanded `-x` and `-F` as well.

File: snapcraft/commands/daemon.start (hunk 2 above)
Comment (simondeziel) [reply snapcraft/commands/daemon.start:622]: The `server_pid 1293` is relative to the container's PID NS which is why it doesn't match the host's view.

Comment (simondeziel) [reply snapcraft/commands/daemon.start:622]: As to why not `--uid` instead of `--euid` it shouldn't really matter here as they should both be `0`. The distinction becomes relevant when using setuid binaries.

Comment (simondeziel) [reply snapcraft/commands/daemon.start:622]: Also, in other scripts we use `id -u` to ensure the EUID is `0` so this should be more consistent.

File: snapcraft/commands/daemon.start [hunk 3]
```diff
@@ -625,12 +625,11 @@ for pid in $(pgrep -f "lxd(\.debug)? --logfile"); do
 
     # Confirm the PID found by pgrep is indeed ours and not one from
     # a nested instance also running LXD or a recycled PID
-    rm -f "/var/snap/lxd/common/lxd/.validate"
-    touch "/proc/${pid}/root/var/snap/lxd/common/lxd/.validate" 2>/dev/null || true
-    if [ -e "/var/snap/lxd/common/lxd/.validate" ]; then
+    echo "$$" > "/proc/${pid}/root/var/snap/lxd/common/lxd/.validate" 2>/dev/null || true
```
Comment (tomponline) [inline snapcraft/commands/daemon.start:633]: I'm not really following why this change is needed (leaving aside the microoptimisation for now too).

What issue is this fixing? Have we seen it occur? 

I'm confused why a nested LXD would be able to write to the /var/snap/lxd/common/lxd directory of its parent?

**Comment by tomponline** [comment]
@simondeziel it'd be interesting to know the context of the precursor discussions you may have had with @mihalicyn about this (he's referenced in the commits) as im not really aware of the overall reason for this PR.

Adding a PR description would likely help to add context here.

File: snapcraft/commands/daemon.start (hunk 3 above)
Comment (simondeziel) [reply snapcraft/commands/daemon.start:633]: It's not fixing any known/observed issue more a theoretical/potential one.

This came about after I asked Aleks if it was OK to replace the touch+rm by just checking for `"/proc/${pid}/root/var/snap/lxd/common/lxd/server.key`'s presence.

Aleks quickly pointed out that would defeat the purpose of that check which is to make sure we are dealing with a LXD PID that belongs to our PID and mount NS. That's because any nested LXD PID we'd get from that `pgrep` would also lead us to a `server.key` but in a different mount NS. Hence the writing to it was needed.

I then observed that if the nested LXD was also going through that same validation at the same time, it could also race and create the `.validate` file to causing problems.

To which Aleks replied: "write and compare random data to eliminate the race".

**Comment by simondeziel** [comment]
@tomponline I've updated the comment based on our discussion to better explain that the `.validate` dance is to rule out *priv* nested LXD by making sure the PID in question doesn't use the same mount namespace as "ours".

I've also reintroduced the removal of that `.validate` file but differently to avoid "leaking" some of them on inside directories of "failed" candidate LXD.

File: snapcraft/commands/daemon.start (hunk 3 above)
Comment (simondeziel) [reply snapcraft/commands/daemon.start:633]: The micro-optimization was dropped.

**Review by tomponline — APPROVED** [review]
